### ISSUE TITLE
Compile with flag `HAVE_NFTSET`

### ISSUE THREAD

**Issue by arakis** [pi-hole/FTL#1508]
(Because community requires trust Level 1, I place it here - hopefully you will read it)

Please consider defining `HAVE_NFTSET`:
https://github.com/pi-hole/FTL/blob/6075a70accdd16a9647a986fce8af51dce03a66f/src/dnsmasq/config.h#L122

Latest FTL/dnsmasq (2.87+) contains a new feature "nftset". There are several cases where `nftsets` are required (regardless of pihole).
For using this feature, the flag `HAVE_NFTSET` needs to be set. This requires at least one dependency: `libnfttables`.
This flag should not harm, even the upcoming Debian "bookworm" is compiling dnsmasq with that flag.

The problem is, ftl replaces dnsmasq, but without that feature. So I cannot use ftl as dnsmaq replacment.

There are four possible workarounds, but all of them are not optimal:
- Fall back to iptables (it has "ipset"). Since my server is using nftables only, this should not mixed up. And IPSET cannot used with nftables.
- Installing pihole as docker, and use ftl as upstream for dnsmasq. Problem: It will not see any client IP. This is my current setup.
- Using macvtap. This will make things much more complicated.
- Running docker as host network. Seems and solid idea, but i have trouble to with port collision (port 53 & 80). Did not found out, how I can change DNS AND HTTP port.

By the way, `HAVE_IPSET` is already defined - but it works only with iptables, nftables cannot read IPSETs.

**Comment by DL6ER** [comment]
You've already seen why we don't do this by default, to keep dependencies low. Actually, we keep it at the bare minimum to avoid version incompatibilities as we'd have to be pretty clear to compile FTL against the *lowest* version available across all the supported systems. Plus, Pi-hole runs on dozens of systems that are not "officially supported", however, adding the `libnfttables` dependency (it is actually `libnftables`, `libnftnl.so`, and `libnfnetlink`, possibly also `libnetfilter_conntrack` that'd be needed) could break Pi-hole on many of these systems.

Debian resolves this issue easily by adding the libraries simply as dependencies and compiling a dedicated binary for every release (so they know library versions match). In contrast, Pi-hole's FTL is designed to run *one* binary both horizontally (distributions) and vertically (releases) at the same time.

TL;DR: I don't have a fixed opinion in this yet, we can still discuss. However, I'm not very much in favor of adding a feature that will (probably) be used by only a few and, on the other hand, probably break compatibility for hundreds or thousands of users running Pi-hole happily on a non-/at-the-edge-supported operating system.

Also pinging @dschaper about his opinion.

**Comment by arakis** [comment]
What you say make sense. 

What about making DNS-Port configurable in the Docker-Installation, like `WEB_PORT=80` for http? For example introducing `FTL_DNS_PORT=53` (existing `FTLPORT=4711` has other purpose, it seems).

In this case, it should not collide anymore with the 0.0.0.0:53 binding of original dnsmasq, when running pihole as docker in host network mode.

**Comment by PromoFaux** [comment]
`dnsmasq` (and thefore pihole-FTL - correct me if I'm wrong @DL6ER) has [`--port`](https://thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html#:~:text=the%20version%20number.-,%2Dp%2C%20%2D%2Dport%3D%3Cport%3E,-Listen%20on%20%3Cport).

Host mount the `/etc/dnsmasq.d` directory and create a file called `01-port.conf`, containing only the line `port=5353` (or whatever port you want), and then restart pihole-FTL, or the container.

```
root@docker-pihole:/etc/dnsmasq.d# cat 01-port.conf
port=5353
root@docker-pihole:/etc/dnsmasq.d# netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:4711          0.0.0.0:*               LISTEN      782/pihole-FTL
tcp        0      0 127.0.0.11:36616        0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:5353            0.0.0.0:*               LISTEN      782/pihole-FTL
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      -
tcp6       0      0 :::5353                 :::*                    LISTEN      782/pihole-FTL
tcp6       0      0 :::80                   :::*                    LISTEN      -
udp        0      0 127.0.0.11:35099        0.0.0.0:*                           -
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           782/pihole-FTL
udp6       0      0 :::5353                 :::*                                782/pihole-FTL
```

**Comment by arakis** [comment]
Exact. I already tried that and will ran into that error, i described here in my comment:
https://github.com/pi-hole/docker-pi-hole/issues/1240#issuecomment-1359065177

**Comment by DL6ER** [comment]
@arakis Note that `port` can be specified only once, do you have it multiple times in your container?
```
grep -R "port=" /etc/dnsmasq.*
```
If so, a fatal error is expected during start as there is no way FTL (resp. `dnsmasq`) can know which of the conflicting setting the one is you actually want.

**Comment by arakis** [comment]
I got it working. Accidentally, I wrote `port 5353` instead of `port=5353` 🤦

**Comment by PromoFaux** [comment]
I was about to say, I just made the change in a container and it survived the container restart.

**Comment by PromoFaux** [comment]
We can look at adding it as an environment variable on the docker side of things though. 

At the moment, you will probably see that your containers stauts is `unhealthy`, because the healthcheck does not know about the port change:

https://github.com/pi-hole/docker-pi-hole/blob/6566662bc3a2829b3d940c65beeab276f23c3edc/src/Dockerfile#L38

**Comment by arakis** [comment]
Ah, this little details explains, why startup timed out, but after setting the new Port also in my local dnsmasq, it's working.

My DNS-Setup is:
- DHCP via dhcpd
- dnsmasq on Debian, using pihole as upstream DNS Server
- pihole listening on Port 5353, configured as --net=host.

Stupidly, i see as client-IP still localhost. But this makes sense and is out of scope here. (the solution would be, to switch the order of pihole and native dnsmasq).

**Comment by PromoFaux** [comment]
I have always found `host` to be a pain in the arse to deal with on docker. Personally I use `macvlan` and give each container it's own IP on my network - don't have to do complicated things to resolve port conflicts then...

**Comment by arakis** [comment]
That's correct, but i was not able to get it running via `macvlan`. Maybe the reason was, `macvlan` does not work well with wifi-adapter (in this case, it's configured as AP).

Anyway, after switching the order of pihole <--> local dnsmasq, all is working fine. I needed also configure `DNSMASQ_LISTENING=all`.
